### PR TITLE
[LoadableByAddress] Fix debug info holes when expanding alloc_stack.

### PR DESCRIPTION
**Pull request by dcci**
We should also skip debug info as they don't really are lowered
to anything real. Add an helper so that we can use it in passes
as well.

@adrian-prantl I'm dealing with a bot failure, but I wanted to put this online so you can take a look and let me know if this is fine. I'm going to add a test & run the tests.

This is the last failure we have before enabling the verifier for debug info, so my plan is to flip the switch today (crossing fingers).

### CODE REVIEW

**Comment by dcci** [comment]
@adrian-prantl now with the correct patch, after coffee.

File: lib/IRGen/LoadableByAddress.cpp [hunk 1]
```diff
@@ -942,11 +942,25 @@ static SILInstruction *createOutlinedCopyCall(SILBuilder &copyBuilder,
   return copy;
 }
 
+// Find the correct debug scope for alloc stack. We want to give to the
+// expanded sequence the correct debug scope so we skip over instructions
+// that aren't lowered to anything real (e.g. debug_value).
+static const SILDebugScope *findAllocStackDebugScope(SILBasicBlock &BB) {
+  auto It = BB.begin();
+  while (It != BB.end()) {
```
Comment (adrian-prantl) [inline lib/IRGen/LoadableByAddress.cpp]: This looks like it could be a range-based-for.

File: lib/IRGen/LoadableByAddress.cpp [hunk 2]
```diff
@@ -942,11 +942,25 @@ static SILInstruction *createOutlinedCopyCall(SILBuilder &copyBuilder,
   return copy;
 }
 
+// Find the correct debug scope for alloc stack. We want to give to the
+// expanded sequence the correct debug scope so we skip over instructions
+// that aren't lowered to anything real (e.g. debug_value).
+static const SILDebugScope *findAllocStackDebugScope(SILBasicBlock &BB) {
```
Comment (adrian-prantl) [inline lib/IRGen/LoadableByAddress.cpp]: I think this utility would make sense in SILBasicBlock.h or even as a member function of SILBasicBlock. It's going to be useful for other passes that have similar issues, too. Perhaps call it `getScopeOfFirstNonMetaInstruction`.

File: lib/IRGen/LoadableByAddress.cpp [hunk 3]
```diff
@@ -942,11 +942,25 @@ static SILInstruction *createOutlinedCopyCall(SILBuilder &copyBuilder,
...
+  return BB.begin()->getDebugScope();
+}
+
 void LoadableStorageAllocation::replaceLoadWithCopyAddr(
     LoadInst *optimizableLoad) {
   SILValue value = optimizableLoad->getOperand();
 
-  SILBuilderWithScope allocBuilder(pass.F->begin()->begin());
+  SILBuilder allocBuilder(pass.F->begin()->begin());
+  allocBuilder.setCurrentDebugScope(findAllocStackDebugScope(*pass.F->begin()));
```
Comment (adrian-prantl) [inline lib/IRGen/LoadableByAddress.cpp]: Wouldn't it be better to add a new constructor to SILBuilderWithScope that has this behavior?

```
/// Creates a new SILBuilder with an insertion point at the beginning of BB and the debug scope from the first non-metainstruction in the BB.
explicit SILBuilderWithScope(SILBasicBlock *BB) {}
```

File: lib/IRGen/LoadableByAddress.cpp [hunk 4]
```diff
@@ -942,11 +942,25 @@ static SILInstruction *createOutlinedCopyCall(SILBuilder &copyBuilder,
   return copy;
 }
 
+// Find the correct debug scope for alloc stack. We want to give to the
+// expanded sequence the correct debug scope so we skip over instructions
+// that aren't lowered to anything real (e.g. debug_value).
+static const SILDebugScope *findAllocStackDebugScope(SILBasicBlock &BB) {
+  auto It = BB.begin();
+  while (It != BB.end()) {
+    if (!isMaintenanceInst(&*It))
```
Comment (adrian-prantl) [inline lib/IRGen/LoadableByAddress.cpp]: Let's rename isMaintenanceInst to isMetaInstruction. This is the same name we are using in MIR.

**Review by adrian-prantl — CHANGES_REQUESTED** [review]

**Comment by dcci** [comment]
@swift-ci please test

**Comment by dcci** [comment]
@adrian-prantl addressed all your comments. Please take another look.

File: include/swift/SIL/DebugUtils.h [hunk 5]
```diff
@@ -49,9 +49,10 @@ inline bool isDebugInst(SILInstruction *Inst) {
   return isa<DebugValueInst>(Inst) || isa<DebugValueAddrInst>(Inst);
 }
 
-/// Returns true if the instruction \p Inst is a maintenance instructions which
-/// is relevant for debug information and does not get lowered to an instruction.
-inline bool isMaintenanceInst(SILInstruction *Inst) {
+/// Returns true if the instruction \p Inst is a meta instruction which
+/// is relevant for debug information and does not get lowered to a real
+/// instruction.
+inline bool isMetaInstruction(SILInstruction *Inst) {
```
Comment (adrian-prantl) [inline include/swift/SIL/DebugUtils.h]: Can you make this a method of SILInstruction?
I also believe that every SILInstruction that implements getVarInfo() should be included in this list (DebugValueAddr, AllocBox, ...) .

**Review by adrian-prantl — CHANGES_REQUESTED** [review]

**Comment by dcci** [comment]
Sure!

**Comment by dcci** [comment]
@swift-ci please test and merge

**Comment by dcci** [comment]
@swift-ci please test and merge

**Review by adrian-prantl — APPROVED** [review]

**Pull request merged**